### DISCUSSION TITLE
IconLayer: How to apply opacity to the layer as a whole, rather than each individual icon

### DISCUSSION THREAD

**Discussion by tech-meppem** [visgl/deck.gl#5599] [Q&A]
I would like to know if it's possible to apply opacity to the IconLayer as a whole, rather than each individual icon.
What I mean by this is that when using either getColor, or the opacity parameters, these are passed to render each icon.
This means that where the icons overlap, the opacities stack.

Here are some mock up images of what I want to do:
With opacity stacking (how it works currently):
![image (4)](https://user-images.githubusercontent.com/54634811/111983129-b1d9c800-8b01-11eb-82ae-b73b20d690e1.png)

Without opacity stacking (what I want to do):
![image (3)](https://user-images.githubusercontent.com/54634811/111983124-b1413180-8b01-11eb-9c3f-5b7764646332.png)

So the way to to this would be to make all the icons opaque, then apply an opacity to the layer as a whole. However, as I previously mentioned, using the opacity option in the layer just passes it to each icon individually, leading to the stacking opacity.

Any help or advice would be greatly appreciated!

Edit: I am using raw JS due to legacy project, non-react.

**Comment by ibgreen** [comment]
All layers have an `opacity` prop. Have you tried it?

**Reply by tech-meppem** [reply]
> All layers have an `opacity` prop. Have you tried it?

Yes I have. I mentioned it in the text, that it just applies the opacity to each icon individually, and doesn't apply to the layer as a whole.

**Reply by kylebarron** [reply]
Have you tried setting the GL blending function? https://github.com/visgl/deck.gl/discussions/5450#discussioncomment-355534

**Reply by tech-meppem** [reply]
> Have you tried setting the GL blending function? [#5450 (comment)](https://github.com/visgl/deck.gl/discussions/5450#discussioncomment-355534)

No, I haven't tried that. How would I do that in plain JS though, without using imports?

Edit: I copied the constants values from luma github, and that didn't really work as expected:
![image](https://user-images.githubusercontent.com/54634811/112016949-12c6c780-8b25-11eb-825d-6469168d6986.png)

**Reply by kylebarron** [reply]
Well if you copied that directly, it wouldn't work because it uses `blendEquation: GL.FUNC_ADD` which is the same as the default I think. My point was that you should look at the GL documentation for other possible blend functions, maybe something like "maximum".

Also there's a bundling option with luma.gl's `GL` export to change to literal values on bundling.

**Reply by tech-meppem** [reply]
@kylebarron With some tweaking, I'm half way there.
I managed to get it like this:
![image](https://user-images.githubusercontent.com/54634811/112031599-6f30e380-8b33-11eb-89bf-582ff6b7027a.png)

So they no longer stack alpha, but the icons have a box around them, which is meant to be transparent.

But thanks for pointing me in the right direction.
